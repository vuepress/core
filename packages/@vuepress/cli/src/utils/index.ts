export * from './allowTs'
export * from './esbuildUtils'
export * from './resolveAppConfig'
export * from './resolveCliAppConfig'
