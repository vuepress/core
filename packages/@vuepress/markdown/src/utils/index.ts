export * from './resolveHeadersFromTokens'
export * from './resolveTitleFromToken'
export * from './vuepressSlugify'
