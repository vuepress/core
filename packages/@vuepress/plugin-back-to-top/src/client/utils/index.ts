export * from './getScrollTop'
export * from './scrollToTop'
