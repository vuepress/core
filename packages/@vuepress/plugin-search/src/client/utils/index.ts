export * from './isKeyMatched'
export * from './isQueryMatched'
