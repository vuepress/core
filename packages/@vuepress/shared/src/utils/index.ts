export { isArray, isFunction, isPromise, isString } from '@vue/shared'

export * from './dedupeHead'
export * from './ensureLeadingSlash'
export * from './ensureEndingSlash'
export * from './formatDateString'
export * from './htmlEscape'
export * from './htmlUnescape'
export * from './isLinkExternal'
export * from './isLinkFtp'
export * from './isLinkHttp'
export * from './isLinkMailto'
export * from './isLinkTel'
export * from './isPlainObject'
export * from './removeEndingSlash'
export * from './removeLeadingSlash'
export * from './resolveHeadIdentifier'
export * from './resolveLocalePath'
export * from './resolveRoutePathFromUrl'
