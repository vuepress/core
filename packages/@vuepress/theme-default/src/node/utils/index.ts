export * from './assignDefaultLocaleOptions'
export * from './resolveContainerPluginOptions'
