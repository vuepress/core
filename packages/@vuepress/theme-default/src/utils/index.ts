export * from './resolveEditLink'
export * from './resolveRepoType'
export * from './scrollWaiter'
