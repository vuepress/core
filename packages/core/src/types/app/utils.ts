/**
 * Directory util function
 */
export type AppDirFunction = (...args: string[]) => string

/**
 * Directory utils
 */
export interface AppDir {
  cache: AppDirFunction
  temp: AppDirFunction
  source: AppDirFunction
  dest: AppDirFunction
  public: AppDirFunction
  client: AppDirFunction
}

/**
 * Environment flags
 */
export interface AppEnv {
  /**
   * Is running in build mode or not
   */
  isBuild: boolean

  /**
   * Is running in dev mode or not
   */
  isDev: boolean

  /**
   * Is debug mode enabled or not
   */
  isDebug: boolean
}

/**
 * Write temp file util
 */
export type AppWriteTemp = (file: string, content: string) => Promise<string>
