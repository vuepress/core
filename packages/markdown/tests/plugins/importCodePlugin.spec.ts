import { fs, path } from '@vuepress/utils'
import MarkdownIt from 'markdown-it'
import { describe, expect, it } from 'vitest'
import { codePlugin, importCodePlugin } from '../../src/index.js'
import type { MarkdownEnv } from '../../src/index.js'

const jsFixturePathRelative = '../__fixtures__/importCode.js'
const mdFixturePathRelative = '../__fixtures__/importCode.md'
const jsFixturePath = path.resolve(__dirname, jsFixturePathRelative)
const mdFixturePath = path.resolve(__dirname, mdFixturePathRelative)
const jsFixtureContent = fs.readFileSync(jsFixturePath).toString()
const mdFixtureContent = fs.readFileSync(mdFixturePath).toString()

describe('@vuepress/markdown > plugins > importCodePlugin', () => {
  it('should not be parsed as import code syntax', () => {
    const source = [
      '@[cod',
      '@[code',
      '@[code]',
      '@[code](./foo.js',
      '@[code](/path/to/foo.js',
      '@[coda](/path/to/foo.js',
    ]

    const md = MarkdownIt().use(importCodePlugin)
    const env: MarkdownEnv = {
      filePath: __filename,
    }
    const rendered = md.render(source.join('\n\n'), env)

    expect(rendered).toEqual(
      source.map((item) => `<p>${item}</p>`).join('\n') + '\n'
    )
    expect(env.importedFiles).toBeUndefined()
  })

  describe('lines range', () => {
    it('should import all lines', () => {
      const source = `\
@[code](${jsFixturePathRelative})
@[code](${mdFixturePathRelative})
`

      const expected = `\
<pre><code class="language-js">\
${jsFixtureContent}\
</code></pre>
<pre><code class="language-md">\
${mdFixtureContent}\
</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual([jsFixturePath, mdFixturePath])
    })

    it('should import partial lines', () => {
      const source = `\
@[code{1-2}](${jsFixturePathRelative})
@[code{1-}](${jsFixturePathRelative})
@[code{4-5}](${mdFixturePathRelative})
@[code{-5}](${mdFixturePathRelative})
`

      const expected = `\
<pre><code class="language-js">\
${jsFixtureContent.split('\n').slice(0, 2).join('\n').replace(/\n?$/, '\n')}\
</code></pre>
<pre><code class="language-js">\
${jsFixtureContent.split('\n').slice(0).join('\n').replace(/\n?$/, '\n')}\
</code></pre>
<pre><code class="language-md">\
${mdFixtureContent.split('\n').slice(3, 5).join('\n').replace(/\n?$/, '\n')}\
</code></pre>
<pre><code class="language-md">\
${mdFixtureContent.split('\n').slice(0, 5).join('\n').replace(/\n?$/, '\n')}\
</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual([
        jsFixturePath,
        jsFixturePath,
        mdFixturePath,
        mdFixturePath,
      ])
    })
  })

  describe('code language', () => {
    it('should use user defined language', () => {
      const source = `\
@[code](/foo.js)
@[code ts](/bar.md)
`

      const expected = `\
<pre><code class="language-js">File not found</code></pre>
<pre><code class="language-ts">File not found</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual(['/foo.js', '/bar.md'])
    })

    it('should use file ext as fallback language', () => {
      const source = `\
@[code](/foo.js)
@[code](/bar.md)
`
      const expected = `\
<pre><code class="language-js">File not found</code></pre>
<pre><code class="language-md">File not found</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual(['/foo.js', '/bar.md'])
    })
  })

  describe('path resolving', () => {
    it('should resolve relative path according to filePath', () => {
      const source = `\
@[code](/foo.js)
@[code](./bar.js)
`
      const expected = `\
<pre><code class="language-js">File not found</code></pre>
<pre><code class="language-js">File not found</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual([
        '/foo.js',
        path.resolve(__dirname, './bar.js'),
      ])
    })

    it('should not resolve relative path if filePath is not provided', () => {
      const source = `\
@[code](/foo.js)
@[code](./bar.js)
`
      const expected = `\
<pre><code class="language-js">File not found</code></pre>
<pre><code class="language-js">Error when resolving path</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: null,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual(['/foo.js'])
    })

    it('should handle import path correctly', () => {
      const source = `\
@[code](@fixtures/importCode.js)
`
      const expected = `\
<pre><code class="language-js">\
${jsFixtureContent}\
</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin, {
        handleImportPath: (str: string): string =>
          str.replace(/^@fixtures/, path.resolve(__dirname, '../__fixtures__')),
      })
      const env: MarkdownEnv = {
        filePath: null,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual([jsFixturePath])
    })
  })

  describe('compatibility with other markdown syntax', () => {
    it('should terminate paragraph', () => {
      const source = `\
foo
@[code](/path/to/foo.js)
`
      const expected = `\
<p>foo</p>
<pre><code class="language-js">File not found</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual(['/path/to/foo.js'])
    })

    it('should terminate blockquote', () => {
      const source = `\
> foo
@[code](/path/to/foo.js)
`
      const expected = `\
<blockquote>
<p>foo</p>
</blockquote>
<pre><code class="language-js">File not found</code></pre>
`

      const md = MarkdownIt().use(importCodePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toEqual(expected)
      expect(env.importedFiles).toEqual(['/path/to/foo.js'])
    })
  })

  describe('compatibility with codePlugin', () => {
    it('should work with syntax supported by codePlugin', () => {
      const source = `\
@[code js{1,3-4}](${jsFixturePathRelative})
@[code md:no-line-numbers](${mdFixturePathRelative})
`

      const md = MarkdownIt().use(importCodePlugin).use(codePlugin)
      const env: MarkdownEnv = {
        filePath: __filename,
      }
      const rendered = md.render(source, env)

      expect(rendered).toMatchSnapshot()
      expect(env.importedFiles).toEqual([jsFixturePath, mdFixturePath])
    })
  })
})
