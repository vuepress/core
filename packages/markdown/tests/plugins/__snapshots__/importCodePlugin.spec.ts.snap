// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@vuepress/markdown > plugins > importCodePlugin > compatibility with codePlugin > should work with syntax supported by codePlugin 1`] = `
"<div class=\\"language-javascript line-numbers-mode\\" data-ext=\\"js\\"><pre v-pre class=\\"language-javascript\\"><code>const msg = 'hello from js'

console.log(msg)

console.log('foo bar')
</code></pre><div class=\\"highlight-lines\\"><div class=\\"highlight-line\\">&nbsp;</div><br><div class=\\"highlight-line\\">&nbsp;</div><div class=\\"highlight-line\\">&nbsp;</div><br></div><div class=\\"line-numbers\\" aria-hidden=\\"true\\"><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div><div class=\\"line-number\\"></div></div></div><div class=\\"language-markdown\\" data-ext=\\"md\\"><pre v-pre class=\\"language-markdown\\"><code># msg

hello from md

## foo bar
</code></pre></div>"
`;
